ruby/computer_science/time_complexity.md

```diff
@@ -0,0 +1,306 @@
+
+You've probably realised the answer is it depends. If you enter 10 like we did before the number of steps is 33 but if you enter another number then the number of steps changes. There is no concrete number we can use to measure the efficiency of our code because it changes based on an external input.
+
+So what we really want to be able to measure is how the number of steps of our algorithm changes when the data changes. This helps us answer the question of whether the code we write will scale.
+
+To do that we need to delve into a new concept. Asymptotic Notations and, in particular, Big O.
+
+### Asymptotic Notations
+
+Simply put, Asymptotic Notations are used to describe the running time of an algorithm. Because an algorithm's running time can differ depending on the input there are several notations that measure that running time in different ways. The 3 most common are as follows:
+
+* Big O Notation - represents the upper bound of an algorithm. This means the worst case scenario for how the algorithm will perform.
+* Omega Notation - represents the lower bound of an algorithm. This is the best case scenario.
+* Theta Notation - represents both the upper bound and lower bound and therefore analyses the average case complexity of an algorithm.
+
+Big O is the one you'll most commonly see referenced because you need to be sure the worst case scenario for any code you write is scalable as the inputs grow in your application.
+
+It's also worth noting that the Notations given below for Big O also apply to Omega and Theta notations. The differences are in how they look to measure the efficiency of the algorithm and therefore which Notation should apply. This should become clearer as you read on.
+
+### What is Big O?
+
+Big O gives us a consistent way to measure the efficiency of an algorithm. It gives us a measurement for the time it takes for an algorithm to run as the input grows so that you can directly compare the performance of two algorithms and pick the best one.
+
+Big O is not a piece of code you can put your algorithm into and it tells you how efficient it is. You will need to measure how the number of steps changes as the data grows and using this you can apply a Big O Notation to it and measure it against other algorithms. In many cases you'll be using a data structure in which the ways you interact with it are well known, and in that case it's easier to judge how it will scale as the input changes.
+
+Firstly we'll summarise the Big O Notations and then provide a little more context for each one. The reading materials will dive into greater detail.
+
+#### Big O Notation
+
+The Big O Notations in the order of speed from fastest to slowest are:
+
+- `O(1)` - Constant Complexity
+- `O(log N)` - Logarithmic Complexity
+- `O(N)` - Linear Complexity
+- `O(N log N)` - N x log N Complexity
+- `O(&#178;)` - Quadratic Complexity
+- `O(&#179;)` - Cubic Complexity
+- `O(2&#8319;)` - Exponential Complexity
+- `O(N!)` - Factorial Complexity
```

Comment:
Not sure if it's only the preview tool, but these don't render as expected: 
![image](https://user-images.githubusercontent.com/85733202/204096191-0f60fc4c-dae7-4336-ad5b-c7a5e45b93ad.png)